### PR TITLE
[PHP] Updates to keep up with PHP 8

### PR DESCRIPTION
**Pull request by Ayesh**
At the moment, highlight.js is a bit behind the major changes in PHP 8.0 and upcoming syntax in PHP 8.1. 
 - PHP 8.0 adds a new [`WeakMap` built-in class](https://php.watch/versions/8.0/weakmap) class.
 - Add `trait` to list of class-like naming patterns. This was supported since PHP 5.4.
 - PHP 8.1 adds support for Enums with [`enum` syntax](https://php.watch/versions/8.1/enums#enum-syntax). Adding `enum` keyword.
 - PHP 8.0 adds a new built-in interface called [`Stringable`](https://php.watch/versions/8.0/stringable).
 - PHP 8.0 match expression throws an [`UnhandledMatchError` exception](https://php.watch/versions/8.0/match-expression#UnhandledMatchError)

This PR updates the PHP keywords list to with these changes. Please note that PHP 8.1 is not yet released, and the Enum syntax is not merged to PHP source yet. The voting is in favor, and it will likely be merged on after the [vote](https://wiki.php.net/rfc/enumerations) ends next week. 

### Changes
 - Add `WeakMap`, `Stringable`, `UnhandledMatchError` to built-in class and interface list. 
 - Update keywords list to include `mixed`. PHP also supports `static` keyword as return/paremter types, but `static` is already in the keywords list. 
 - Add `trait` and `enum` to class-like name matching list.

### Checklist
- [x] Added markup tests, or they don't apply here because...
- [x] Updated the changelog at `CHANGES.md`
- [x] Added myself to `AUTHORS.txt`, under Contributors

### Screenshots

image

### CODE REVIEW

File: test/detect/php/default.txt
Comment (joshgoebel) [reply test/detect/php/default.txt:51]: IE: See the failing tests "AssertionError: default.txt should be detected as php, but was zephir".

This is likely because you added code here that "looks like" Zephir (which is very close to PHP).

I'd suggest reverting this part of your changes and seeing if that makes tests green.

**Review by joshgoebel — COMMENTED** [review]
> Please note that PHP 8.1 is not yet released, and the Enum syntax is not merged to PHP source yet. The voting is in favor, and it will likely be merged on after the vote ends next week.

So then we probably hold this PR at least until the end of the voting.

---

Looks like you still have a problem with the markup tests failing... is it possible you've added code that hits an "illegal"? (which short circuits all the highlighting?)  That's what it looked like just at a glance...

Are your tests passing locally?

**Comment by joshgoebel** [comment]
Thanks for the great write-up and clean commit history.  Very nice.

File: SUPPORTED_LANGUAGES.md
Comment (joshgoebel) [inline SUPPORTED_LANGUAGES.md:139]: Curious: Does PHP use `php8` or `php5` as file extensions?   I'm not sure why we need adding version numbers here... (I'm not questing THIS particular add, just the pattern)  

We don't do anything different for different versions of PHP. I'm thinking with version 11 all these need to disappear and just have the single "php"? Do you know anything that would suggest otherwise?

Comment (Ayesh) [reply SUPPORTED_LANGUAGES.md:139]: I'd be hugely in favor to get rid of all version-specific extensions.

As far as I remember, the `php5` extension/mime-type was used in cPanel default configuration options. I don't think anybody continues this pattern anymore. There is no PHP 6 at all either. I think this list was just kept updated by incrementing just this number, but not dropping anything in the name of backwards compatibility?

File: test/detect/php/default.txt
Comment (Ayesh) [reply test/detect/php/default.txt:51]: I will try this, thank you. I used that particular snippet because it has the new `mixed` type in it, but I think at least for the `default.txt` file, the simpler, the better.

**Comment by Ayesh** [comment]
Thanks a lot for the review @joshgoebel - I'm so grateful for these insights. I will work on the failing tests. They fail locally (node 15.2) for me as well. I'm not 100% sure what keeps them from failing, but I will take a thorough look.

File: SUPPORTED_LANGUAGES.md
Comment (joshgoebel) [reply SUPPORTED_LANGUAGES.md:139]: Before my time.  This is a potential breaking change though, so we're stuck with it for now.  But I will add it to the list for v11 changes.

**Comment by joshgoebel** [comment]
Try commenting out all the `illegal` keys and see if it starts working... if so that's a clue.

File: test/detect/php/default.txt
Comment (joshgoebel) [reply test/detect/php/default.txt:51]: `default.txt` should in practice be a "clear, representative sample" of a language... no more, no less.  So perhaps they should change indeed slowly over time, but the current system makes it frustrating to do so.

The current auto-detect system is all relative to other languages vs being more yes/no... if it was a more isolated and binary decision such as "this is PHP" or "this is not" these tests would make a lot more sense.

Our sample size is really too small for what we're attempting to do with it anyways - so it's become a very artificial thing.

**Comment by Ayesh** [comment]
Thank you, tinkering with `illegal` key did the trick! There was a `[:($"]` regex there, and it was ruling out the potential `enum Foo: string` syntax because of the `:` sign. I relaxed the illegal regex to allow `:` characters. Triggering CI because it passed locally. 

I will update with a negative look behind to tune the regex there.

**Comment by joshgoebel** [comment]
> I will update with a negative look behind to tune the regex there.

Can't use negative look-behinds, they aren't supported by all browsers yet.

**Comment by Ayesh** [comment]
I understand, thank you.

I can't think about a way to make the `illegal` rule more strict. I think allowing `:` character is worth the relaxed `illegal` rule to allow `enum Foo: string {}` syntax. 

> So then we probably hold this PR at least until the end of the voting.
Yes. I will remove the draft status from this PR on Feb 17, that's when the vote should end and code gets officially merged. 

Thanks a lot for shedding the light throughout this PR @joshgoebel.

File: src/languages/php.js
Comment (joshgoebel) [inline src/languages/php.js:159]: Try using `variants` here, something like:

```
variants: [
  { beginKeywords: "enum", illegal: [reduced set] },
  { beginKeywordS: "class interface trait", illegal: [full set] }
]
```

It's a shortcut to writing all new rules (which is what the mode compiler does internally).. All the other keys (from the top) are inherited.

**Comment by Ayesh** [comment]
Brilliant, that was great idea. Tests should pass now too. Thank you.

**Comment by joshgoebel** [comment]
You seem to know your PHP, perhaps you could add your thoughts to: https://github.com/highlightjs/highlight.js/issues/2996 if you have time.

**Comment by Ayesh** [comment]
@joshgoebel Spot on with the `beginKeywordS`. I force-pushed with it fixed.

**Review by joshgoebel — APPROVED** [review]
Pending making sure PHP 8.1 goes as planned.

**Comment by Ayesh** [comment]
I added a new commit that expands our number detection. It now supports all numerals supported up to PHP 8.1.
 - Decimal numbers
 - Floats
 - Binary numbers (`0b1100` and `0B1100`).
 - Hex number support (`0xAFAF` and `0XAFAF`).
 - Octal number support (`0777`)
 - [New `0o` and `0O` prefix support for Octals](https://php.watch/versions/8.1/explicit-octal-notation) (`0o777` and `0O777`) (Already accepted and merged to upcoming PHP 8.1)
 - Scientific notation (`7E-10` and `1.2e3`)
 - [Underscore number separator](https://php.watch/versions/7.4/underscore_numeric_separator) (Already implemented in PHP 7.4)


image

Thank you.

File: src/languages/php.js
Comment (joshgoebel) [inline src/languages/php.js:66]: Nice.

Please expand this into multiple variants for each variant of number vs a huge regex that tends to be impossible to read and maintain over time.

Comment (Ayesh) [reply src/languages/php.js]: Thank you. It's now expanded into four variants - binary, octal, hex, and decimal.

**Comment by Ayesh** [comment]
I think this PR will creep up with more changes, so I will wait for the Enum vote to finish on Feb 17, and then take out this PR from the draft state. Once these changes are made, I will see to open another with the only two missing syntax changes:

So far, the two other features we could improve are:
 - Attributes (`#[Foo(Bar)]`): They are currently matched as a comment because PHP also follows hash-comment syntax. 
 - Named parameters: (`call_foo(foo: 1, bar: 2)`). I could get attributes to work, but I will do some more testing with named parameters.

**Comment by joshgoebel** [comment]
> I think this PR will creep up with more changes

Personally I'd rather see this frozen and then another PR opened (it can be on top of this commit history, and rebased later) - but I'm also trying to be flexible since you're one of the cleanest committers I've ever seen here. :-)  Perhaps the real issue is my rush to review things, lol.  Don't get a lot of draft PRs.

> Attributes (#[Foo(Bar)])

That would be a great and easy add I think... would these be `meta` then?

> Named parameters: (call_foo(foo: 1, bar: 2)).

This involves detecting `functionDispatch(` style patterns, most likely, yes?  If so we should highlight the dispatch also (function name), though there is some question on what class we should see. See #2500, etc.  But if you just slap something on it I can sort out the proper class afterwards.

**Comment by Ayesh** [comment]
Hi @joshgoebel - thanks for the help with attributes and named parameters. You are right, I think it's good to go with "meta" for Attributes. Rust seems to get same treatment too.

---

So far, the only change that is not yet landed on PHP is Enums. All other features are already released or merged. I rebased off #3000, and applied all other patches except a couple related to Enums. I will send a separate PRs for Enums, Attributes, and Named parameters. 

Thank you.

**Comment by Ayesh** [comment]
Taking this PR out of draft state, because all changes are already landed in PHP, and all changes except octal numerals are working in the already released PHP 8.0. 
Fingers crossed :)

**Review by joshgoebel — APPROVED** [review]

**Comment by joshgoebel** [comment]
@Ayesh Thanks so much!

**Comment by Ayesh** [comment]
Yay, thank you.

**Pull request merged**
